Modules/Utils/Sources/PocketCastsUtils/Extensions/JSONCodable.swift

```diff
@@ -0,0 +1,86 @@
+import Foundation
+
+/**
+ This provides some protocols to reduce the boilerplate of encoding and decoding objects to/from JSON.
+ JSONCodable is a typealias of the JSONEncodable and JSONDecodable protocols.
+
+ ## JSONEncodable
+ This inherits from the Encodable protocol and defines a `jsonData` property which
+ by default will encode the object using a basic JSONEncoder() and will return nil on failure.
+
+ ## JSONDecodable
+ This inherits from the Decodable protocol and defines the `encodeObject(type:from)` method which by default will attempt
+ to decode the given `data` to the specified `type` and will return nil if the decoding fails or can't be cast to the `type`.
+
+ ## Example Usage:
+
+ enum HelloWorldEnum: JSONCodable {
+    case one, two, three
+ }
+
+ print(HelloWorldEnum.one.jsonData as? NSData)
+    Outputs -> Optional(<7b226f6e 65223a7b 7d7d>) -> {"one":{}}
+
+ struct HelloWorldStruct: JSONCodable {
+     let one: String
+     let two: HelloWorldEnum
+ }
+
+ print(HelloWorld(one: "one", two: .two).jsonData as? NSData)
+    Outputs -> Optional(<7b227477 6f223a7b 2274776f 223a7b7d 7d2c226f 6e65223a 226f6e65 227d>) -> {"two":{"two":{}},"one":"one"}
+
+ ## UserDefaults
+ This also provides a UserDefaults extension that allows easy reading and writing of
+ encoded values using `set(encodedValue:forKey)` and `jsonObject(type:forKey)`
+
+ The methods behave similar to the default `set(value:forKey)` and `object(forKey:)` methods.
+ */
+
+// MARK: - JSONEncodable
+
+/// A type that can encode itself as JSON data using a JSONEncoder
+public protocol JSONEncodable: Encodable {
```

Comment:
There's any reason why we need an abstraction on top of `Encodable` and `Decodable`?

I think we can highly simplify the logic if we keep using `Codable` in general and I think it increases the readability, we can reduce everything to:

```swift
public extension UserDefaults {
    /// Saves a JSONEncodable object to the UserDefaults for the given key.
    /// Passing nil to this will delete the value.
    func set<Value: Codable>(_ encodedValue: Value?, forKey key: String) {
        setValue(try? JSONEncoder().encode(encodedValue), forKey: key)
    }

    /// Retrieves the JSONDecodable object from the UserDefaults for the given key.
    /// THe value and cast it to the given `type`
    func jsonObject<Value: Decodable>(_ type: Value.Type, forKey: String) -> Value? {
        data(forKey: forKey).flatMap {
            try? JSONDecoder().decode(type, from: $0)
        }
    }
}
```

Then we can make `HeadphoneControlAction` conform to `Codable`.

Reply:
I originally wrote it like this but this also means any object that conforms to Codable will default to being saved as JSON which includes Bool, String, etc...

I felt like doing that was too much overhead for storing simple values since even a bool will be saved as:

```swift
// true
<74727565>
```

Using the abstraction ensures we're only saving the values we need to save as JSON.